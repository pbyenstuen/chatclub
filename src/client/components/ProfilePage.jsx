import React, { useEffect } from "react";
import useLoader from "./useLoader";
import ErrorView from "./ErrorView";

const ProfilePage = ({ api }) => {
    const { data: user, loading, error, reload } = useLoader(async () => await api.auth.getUser());

    if (loading) {
        return <h2>Loading...</h2>
    }

    if (error) {
        return <ErrorView error={error} reload={reload} />;
    }

    return (
        <div id="profile-container">
            <header>
                <h2>Profile</h2>
            </header>
            <div id="profile-card">
                {Object.entries(user).length > 2 ?
                    <>
                        <img src={user.image} alt={user.displayName} />
                        <div>
                            <h4>USERNAME/EMAIL</h4>
                            <p>{user.username}</p>
                            <h4>DISPLAY NAME</h4>
                            <p>{user.displayName}</p>
                            <h4>FIRST NAME</h4>
                            <p>{user.firstName}</p>
                            <h4>LAST NAME</h4>
                            <p>{user.lastName}</p>
                        </div>
                    </>
                    :
                    <div>
                        <h4>USERNAME</h4>
                        <p>{user.username}</p>
                    </div>}
            </div>
        </div>
    )
}

export default ProfilePage;
