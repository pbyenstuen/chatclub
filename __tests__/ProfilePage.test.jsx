import React from "react";
import ReactDOM from "react-dom";
import { act, Simulate } from "react-dom/test-utils";
import { MemoryRouter } from "react-router";
import ProfilePage from "../src/client/components/ProfilePage";

let container;

beforeEach(() => {
    container = document.createElement('div');
    document.body.appendChild(container);
});

afterEach(() => {
    document.body.removeChild(container);
    container = null;
});

const mountComponent = async (component) => {
    await act(async () => {
        await ReactDOM.render(<MemoryRouter>{component}</MemoryRouter>, container);
    });
}

describe("ProfilePage", () => {
    it("can show user info (google)", async () => {
        const api = {
            auth: {
                getUser: () => ({
                    username: "Test",
                    displayName: "Test",
                    firstName: "Test",
                    lastName: "Test",
                    image: "test.png"
                })
            }
        }

        await mountComponent(<ProfilePage api={api} />);

        expect(container.innerHTML).toMatchSnapshot();
        container.querySelectorAll("p").forEach(p => {
            expect(p.textContent).toEqual("Test");
        });
    });

    it("can show user info (local)", async () => {
        const api = {
            auth: {
                getUser: () => ({
                    username: "Test",
                    displayName: "Test",
                })
            }
        }

        await mountComponent(<ProfilePage api={api} />);

        expect(container.querySelector("p").textContent).toEqual("Test");
    });

    it("can show error message", async () => {
        const api = {
            auth: {
                getUser: () => {
                    throw new Error("Error")
                }
            }
        }

        await mountComponent(<ProfilePage api={api} />);

        expect(container.innerHTML).toMatchSnapshot();
        expect(container.querySelector("div").textContent).toEqual("Something went wrong: Error: Error");
    });
});
