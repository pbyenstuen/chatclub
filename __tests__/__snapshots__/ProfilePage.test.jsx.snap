// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfilePage can show error message 1`] = `"<div>Something went wrong: Error: Error</div><button>TRY AGAIN</button>"`;

exports[`ProfilePage can show user info (google) 1`] = `"<div id=\\"profile-container\\"><header><h2>Profile</h2></header><div id=\\"profile-card\\"><img src=\\"test.png\\" alt=\\"Test\\"><div><h4>USERNAME/EMAIL</h4><p>Test</p><h4>DISPLAY NAME</h4><p>Test</p><h4>FIRST NAME</h4><p>Test</p><h4>LAST NAME</h4><p>Test</p></div></div></div>"`;
